### ISSUE TITLE
X11 WSI support for android targets?

### ISSUE THREAD

**Issue by twaik** [ptitSeb/gl4es#435]
Hi. GL4ES can be used on Android devices but there is no X11 support. Can you please port mesa's X11 WSI code ([this](https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/loader/loader_dri3_helper.c) + [this](https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/egl/drivers/dri2/platform_x11_dri3.c))? Or maybe mesa's vulkan X11 WSI code ([this](https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/vulkan/wsi/wsi_common_x11.c))? You can use AHardwareBuffer API for EGLImage and set it as framebuffer/renderbuffer and attach AHardwareBuffer's udmabuf fd to X server with simple `xcb_shm_attach_fd` call and then draw it with `xcb_shm_put_image` call.

Thank you.

**Comment by OpomboMORTO** [comment]
Interestingly, this would cause the framebuffer to be read by the AHardwareBuffer and sending it to x11, correct?

**Comment by twaik** [comment]
Not exactly. It will cause rendering directly to AHardwareBuffer, which will be read by X server during rendering the root window.

**Comment by OpomboMORTO** [comment]
I hope the implementation is soon

**Comment by OpomboMORTO** [comment]
You can try opening another poll to try to get an answer

**Comment by Mateus2022** [comment]
@ptitSeb please add framebuffer and x11 support when using gl4es on termux or termux-x11 platforms in case of android definitions without NOX11

**Comment by ptitSeb** [comment]
> @ptitSeb please add framebuffer and x11 support when using gl4es on termux or termux-x11 platforms in case of android definitions without NOX11

And how do that?
I mean, how do I create a context and all the stuffs needed to emulate X1 when NOX11 is used?

**Comment by twaik** [comment]
You can create wrap all EGL calls and emulate X11 surface. Try to read this. https://github.com/tareksander/termux-gfx-wrapper/tree/main/src/egl

**Comment by GuilhermeDunkly** [comment]
what is NOX11?
